### PR TITLE
`Programming exercises`: Speedup archival

### PR DESCRIPTION
**Pull request by krusche**
Archiving programming exercises as part of a course or exam was done sequentially which can take quite some time, especially in courses with between 10.000 and 80.000 repositories (such as on the TUM production system).

This PR speeds up the process by running the clone and zip operation in 10 parallel threads. In a local test, the archival time of one course was around 5-7x faster with this approach.

#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Steps for Testing

Prerequisites:
- Archive a course (and an exam) with passed end date on the course management detail page
- Download the archive and verify all student repositories are included

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 1]
```diff
@@ -535,20 +537,26 @@ public List<Path> exportStudentRepositories(ProgrammingExercise programmingExerc
...
-            catch (Exception e) {
+            catch (Exception exception) {
                 var error = "Failed to export the student repository with participation: " + participation.getId() + " for programming exercise '" + programmingExercise.getTitle()
                         + "' (id: " + programmingExercise.getId() + ") because the repository couldn't be downloaded. ";
                 exportErrors.add(error);
             }
-        });
+        }, threadPool).toCompletableFuture()).toList().toArray(new CompletableFuture<?>[participations.size()]);
+        // wait until all operations finish
+        CompletableFuture.allOf(futures).join();
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:559]:
```suggestion
        CompletableFuture.allOf(futures).thenAccept((emtpy) -> threadPool.shutdown()).join();
```

**Review by Strohgelaender — APPROVED** [review]
code lgtm

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 2]
```diff
@@ -535,20 +537,26 @@ public List<Path> exportStudentRepositories(ProgrammingExercise programmingExerc
...
                 }
             }
-            catch (Exception e) {
+            catch (Exception exception) {
                 var error = "Failed to export the student repository with participation: " + participation.getId() + " for programming exercise '" + programmingExercise.getTitle()
                         + "' (id: " + programmingExercise.getId() + ") because the repository couldn't be downloaded. ";
                 exportErrors.add(error);
             }
-        });
+        }, threadPool).toCompletableFuture()).toList().toArray(new CompletableFuture<?>[participations.size()]);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:557]: We had a PR earlier that replaced such calls with a zero with array claiming to improve performance and synchronization issues.
See #5885

Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:557]:
```suggestion
        }, threadPool).toCompletableFuture()).toArray(CompletableFuture[]::new);
```

**Pull request merged**